File: TouchNothing/Utilities/SupabaseRPCErrorMapper.swift
import Foundation
import Supabase

enum SupabaseRPCErrorMapper {
    private static let authErrorTokens: [String: SupabaseServiceError] = [
        "nickname_taken": .nicknameTaken,
        "invalid_credentials": .invalidCredentials(),
        "account_locked": .accountLocked,
        "invalid_pin": .invalidPin,
        "invalid_nickname": .invalidNickname,
        "invalid_session_token": .invalidSessionToken,
        "nickname_not_registered": .nicknameNotRegistered
    ]

    static func mapAuthErrorToken(_ token: String?) -> SupabaseServiceError? {
        guard let token else { return nil }
        return authErrorTokens[token]
    }

    static func map(_ error: Error) -> SupabaseServiceError {
        if let postgrestError = error as? PostgrestError {
            return mapMessage(postgrestError.message.lowercased())
        }

        let message = String(describing: error).lowercased()

        if isLikelyEmptyAuthResponse(error) {
            return .serverUpgradeRequired
        }

        return mapMessage(message)
    }

    static func isLikelyEmptyAuthResponse(_ error: Error) -> Bool {
        if error is DecodingError {
            return true
        }

        let message = String(describing: error).lowercased()
        return message.contains("decoding")
            || message.contains("correct format")
            || message.contains("unexpected end of file")
    }

    private static func mapMessage(_ message: String) -> SupabaseServiceError {
        for (token, error) in authErrorTokens where message.contains(token) {
            return error
        }

        if message.contains("function"), message.contains("does not exist") {
            return .serverUpgradeRequired
        }

        return .networkFailure
    }
}

File: app/src/main/java/com/touchnothing/app/data/remote/SupabaseRpcErrorMapper.kt
package com.touchnothing.app.data.remote

object SupabaseRpcErrorMapper {
    private val authErrorTokens = mapOf(
        "nickname_taken" to SupabaseServiceError.NicknameTaken,
        "invalid_credentials" to SupabaseServiceError.InvalidCredentials(),
        "account_locked" to SupabaseServiceError.AccountLocked,
        "invalid_pin" to SupabaseServiceError.InvalidPin,
        "invalid_nickname" to SupabaseServiceError.InvalidNickname,
        "invalid_session_token" to SupabaseServiceError.InvalidSessionToken,
        "nickname_not_registered" to SupabaseServiceError.NicknameNotRegistered,
    )

    fun mapAuthErrorToken(token: String?): SupabaseServiceError? {
        if (token == null) return null
        return authErrorTokens[token]
    }

    fun map(error: Throwable): SupabaseServiceError {
        val message = error.message?.lowercase() ?: error.toString().lowercase()

        if (isLikelyEmptyAuthResponse(error)) {
            return SupabaseServiceError.ServerUpgradeRequired
        }

        return mapMessage(message)
    }

    fun mapMessage(message: String): SupabaseServiceError {
        val lower = message.lowercase()
        authErrorTokens.forEach { (token, serviceError) ->
            if (lower.contains(token)) return serviceError
        }
        if (lower.contains("function") && lower.contains("does not exist")) {
            return SupabaseServiceError.ServerUpgradeRequired
        }
        return SupabaseServiceError.NetworkFailure
    }

    fun isLikelyEmptyAuthResponse(error: Throwable): Boolean {
        val message = error.message?.lowercase() ?: error.toString().lowercase()
        return message.contains("decoding") ||
            message.contains("correct format") ||
            message.contains("unexpected end of file") ||
            message.contains("expected start")
    }
}
